# Assignment 6



## PANDAS 


### Q1- Create a Pandas Data frame from the given data and create a new column “Voter” based on voter age, i.e., if age >18 then voter column should be “Yes” otherwise if age <18 then voter column should be “No”

```python
raw_Data = {'Voter_name': ['Geek1', 'Geek2', 'Geek3', 'Geek4',
'Geek5', 'Geek6', 'Geek7', 'Geek8'],
'Voter_age': [15, 23, 25, 9, 67, 54, 42, np.NaN]}
```

#### Import Libraries

In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings("ignore") # To ignore unnecessary warnings

In [2]:
raw_Data = {'Voter_name': ['Geek1', 'Geek2', 'Geek3', 'Geek4',
'Geek5', 'Geek6', 'Geek7', 'Geek8'],
'Voter_age': [15, 23, 25, 9, 67, 54, 42, np.NaN]}

df = pd.DataFrame(raw_Data) # convert dictionary raw_Data into a Data Frame.

df

Voter_name  Voter_age
0      Geek1       15.0
1      Geek2       23.0
2      Geek3       25.0
3      Geek4        9.0
4      Geek5       67.0
5      Geek6       54.0
6      Geek7       42.0
7      Geek8        NaN

#### Handling Missing Values

In [3]:
df.info() # Check count of not-null values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Voter_name  8 non-null      object 
 1   Voter_age   7 non-null      float64
dtypes: float64(1), object(1)
memory usage: 260.0+ bytes


In [4]:
row_with_NaN = df[df['Voter_age'].isnull()] # finding the row with null values
row_with_NaN

Voter_name  Voter_age
7      Geek8        NaN

#### 1. DROP MISSING VALUE (or) NAN

In [5]:
df_droppedValue = df.dropna() # save the Data Frame after droping null values
df_droppedValue

Voter_name  Voter_age
0      Geek1       15.0
1      Geek2       23.0
2      Geek3       25.0
3      Geek4        9.0
4      Geek5       67.0
5      Geek6       54.0
6      Geek7       42.0

In [6]:
# Create a function for checking out the Voting_age is above 18 or below

def voting_age(age):
    if age > 18:
        return 'Yes'
    else:
        return 'No'

In [7]:
# Creating another column 'Voter' to save Yes or No according to their age
df_droppedValue['Voter'] = df_droppedValue['Voter_age'].apply(voting_age)
df_droppedValue

Voter_name  Voter_age Voter
0      Geek1       15.0    No
1      Geek2       23.0   Yes
2      Geek3       25.0   Yes
3      Geek4        9.0    No
4      Geek5       67.0   Yes
5      Geek6       54.0   Yes
6      Geek7       42.0   Yes

#### 2. Impute Missing Value (or) NaN with the mean()

In [8]:
# coping the data frame in other variable df_mean for experimentation impute with mean() 
df_mean = df
mean_age = df_mean['Voter_age'].mean()

# fill the null value (or) missing value with the median value
df_mean['Voter_age'] = df_mean['Voter_age'].fillna(mean_age)

# Creating another column 'Voter' to save Yes or No according to their age
df_mean['Voter'] = df_mean['Voter_age'].apply(voting_age)
df_mean

Voter_name  Voter_age Voter
0      Geek1  15.000000    No
1      Geek2  23.000000   Yes
2      Geek3  25.000000   Yes
3      Geek4   9.000000    No
4      Geek5  67.000000   Yes
5      Geek6  54.000000   Yes
6      Geek7  42.000000   Yes
7      Geek8  33.571429   Yes

#### 3. Impute Missing Value (or) NaN with median()

In [9]:
# coping the data frame in other variable df_median for experimentation impute with median() 
df_median = df
median_age = df_median['Voter_age'].median()

# fill the null value (or) missing value with the median value
df_median['Voter_age'] = df_median['Voter_age'].fillna(median_age)

# Creating another column 'Voter' to save Yes or No according to their age
df_median['Voter'] = df_median['Voter_age'].apply(voting_age)
df_median

Voter_name  Voter_age Voter
0      Geek1  15.000000    No
1      Geek2  23.000000   Yes
2      Geek3  25.000000   Yes
3      Geek4   9.000000    No
4      Geek5  67.000000   Yes
5      Geek6  54.000000   Yes
6      Geek7  42.000000   Yes
7      Geek8  33.571429   Yes

### Q2 – Create a Pandas Data frame from the given data and collapse First and Last column into one column as Full Name, so the output contains Full Name and Age, then convert column age to index

```python
raw_Data = {'First': ['Manan ', 'Raghav ', 'Sunny '],
'Last': ['Goel', 'Sharma', 'Chawla'],
'Age' : [12, 24, 56]}
```

#### Step 1: Creating DataFrame()

In [10]:
raw_Data = {'First': ['Manan ', 'Raghav ', 'Sunny '],
'Last': ['Goel', 'Sharma', 'Chawla'],
'Age' : [12, 24, 56]}

df = pd.DataFrame(raw_Data)

df

First    Last  Age
0   Manan     Goel   12
1  Raghav   Sharma   24
2   Sunny   Chawla   56

#### Step 2: Checking Null values (or) Missing values

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   First   3 non-null      object
 1   Last    3 non-null      object
 2   Age     3 non-null      int64 
dtypes: int64(1), object(2)
memory usage: 204.0+ bytes


In [12]:
df.isnull().sum()

First    0
Last     0
Age      0
dtype: int64

#### Step 3: Collapse columns 'First' and 'Last' into new column 'Full Name'

In [13]:
df['Full Name'] = df['First'] + ' ' + df['Last']

df

First    Last  Age       Full Name
0   Manan     Goel   12     Manan  Goel
1  Raghav   Sharma   24  Raghav  Sharma
2   Sunny   Chawla   56   Sunny  Chawla

#### Step 4: Drop the columns 'First' and 'Last' (they are of no use)

In [14]:
df.drop(['First', 'Last'], axis=1, inplace=True)
df

Age       Full Name
0   12     Manan  Goel
1   24  Raghav  Sharma
2   56   Sunny  Chawla

#### Step 5: Set column 'Age' as index

In [15]:
df.set_index('Age', inplace=True)

In [16]:
df

Full Name
Age                
12      Manan  Goel
24   Raghav  Sharma
56    Sunny  Chawla

### Q3- Create a Pandas Data frame from the given data -

```python
raw_Data = {'Date':['10/2/2011', '11/2/2011', '12/2/2011', '13/2/2011'],
'Product':['Umbrella', 'Matress', 'Badminton','Shuttle'],
'Price':[1250, 1450, 1550, 400],
'Expense': [ 21525220.653, 31125840.875, 23135428.768, 56245263.942]}
```

In [17]:
raw_Data = {'Date':['10/2/2011', '11/2/2011', '12/2/2011', '13/2/2011'],
'Product':['Umbrella', 'Matress', 'Badminton','Shuttle'],
'Price':[1250, 1450, 1550, 400],
'Expense': [ 21525220.653, 31125840.875, 23135428.768, 56245263.942]}

df = pd.DataFrame(raw_Data)

df

Date    Product  Price       Expense
0  10/2/2011   Umbrella   1250  2.152522e+07
1  11/2/2011    Matress   1450  3.112584e+07
2  12/2/2011  Badminton   1550  2.313543e+07
3  13/2/2011    Shuttle    400  5.624526e+07

#### a)  Add Index as Item1, Item2, Item3, Item4

In [18]:
# Add Index as Item1, Item2, Item3, Item4
df.index = ['Item1', 'Item2', 'Item3', 'Item4']
df

Date    Product  Price       Expense
Item1  10/2/2011   Umbrella   1250  2.152522e+07
Item2  11/2/2011    Matress   1450  3.112584e+07
Item3  12/2/2011  Badminton   1550  2.313543e+07
Item4  13/2/2011    Shuttle    400  5.624526e+07

#### b)  Find the index labels of all items whose ‘Price’ is greater than 1000.

In [19]:
# Index labels of items whose ‘Price’ is greater than 1000
high_price_items = df[df['Price'] > 1000].index

high_price_items

Index(['Item1', 'Item2', 'Item3'], dtype='object')

#### c) Replace products using Map() with respective codes- Umbrella : ‘U’, Matress : 'M', Badminton : 'B', Shuttle: 'S'

In [20]:
# Replace products using Map() with respective codes
product_map = {'Umbrella': 'U', 'Matress': 'M', 'Badminton': 'B', 'Shuttle': 'S'}
df['Product'] = df['Product'].map(product_map)
df

Date Product  Price       Expense
Item1  10/2/2011       U   1250  2.152522e+07
Item2  11/2/2011       M   1450  3.112584e+07
Item3  12/2/2011       B   1550  2.313543e+07
Item4  13/2/2011       S    400  5.624526e+07

#### d) Round off the Expense column values to two decimal places.

In [21]:
# Round off the Expense column values to two decimal places
df['Expense'] = df['Expense'].round(2)
df

Date Product  Price      Expense
Item1  10/2/2011       U   1250  21525220.65
Item2  11/2/2011       M   1450  31125840.88
Item3  12/2/2011       B   1550  23135428.77
Item4  13/2/2011       S    400  56245263.94

#### e) Create a new column called ‘Discounted_Price’ after applying a 10% discount on the existing‘price’ column. (try using lambda function)


In [22]:
# Create a new column called ‘Discounted_Price’ after applying a 10% discount
df['Discounted_Price'] = df['Price'].apply(lambda x: x * 0.9)
df

Date Product  Price      Expense  Discounted_Price
Item1  10/2/2011       U   1250  21525220.65            1125.0
Item2  11/2/2011       M   1450  31125840.88            1305.0
Item3  12/2/2011       B   1550  23135428.77            1395.0
Item4  13/2/2011       S    400  56245263.94             360.0

#### f) Convert the column type of “Date” to datetime format

In [23]:
# Convert the column type of “Date” to datetime format
df['Date'] = pd.to_datetime(df['Date'])
df

Date Product  Price      Expense  Discounted_Price
Item1 2011-10-02       U   1250  21525220.65            1125.0
Item2 2011-11-02       M   1450  31125840.88            1305.0
Item3 2011-12-02       B   1550  23135428.77            1395.0
Item4 2011-02-13       S    400  56245263.94             360.0

#### g) Create a column rank which ranks the products based on the price (one with the highest price will be rank 1).

In [24]:
# Create a column rank which ranks the products based on the price
df['Rank'] = df['Price'].rank(ascending=False).astype(int)
df

Date Product  Price      Expense  Discounted_Price  Rank
Item1 2011-10-02       U   1250  21525220.65            1125.0     3
Item2 2011-11-02       M   1450  31125840.88            1305.0     2
Item3 2011-12-02       B   1550  23135428.77            1395.0     1
Item4 2011-02-13       S    400  56245263.94             360.0     4

# Assignment: Exploring NBA Player Data

Download the nba.csv file containing NBA player data Complete the following tasks using
Python, Pandas, and data visualization libraries:

#### 1. Load Data:
    ● Load the nba.csv data into a Pandas DataFrame.
    ● Display basic information about the DataFrame.
#### 2. Data Cleaning:
    ● Handle missing values by either removing or imputing them.
    ● Remove duplicate rows.
#### 3. Data Transformation:
    ● Create a new column 'BMI' (Body Mass Index) using the formula: BMI = (weight in pounds / (height in inches)^2) * 703.(Assuming a fixed height value of 70 inches (5 feet 10 inches)
#### 4. Exploratory Data Analysis (EDA):
    ● Display summary statistics of the 'age', 'weight', and 'salary' columns.
    ● Calculate the average age, weight, and salary of players in each 'position' category.
#### 5. Data Visualization:
    ● Create a histogram of player ages.
    ● Create a boxplot of player salaries for each 'position'.
    ● Plot a scatter plot of 'age' vs. 'salary' with a different color for each 'position'.
#### 6. Top Players:
    ● Display the top 10 players with the highest salaries.
#### 7. College Analysis:
    ● Determine the top 5 colleges with the most represented players.
#### 8. Position Distribution:
    ● Plot a pie chart to show the distribution of players across different 'positions'.
#### 9. Team Analysis:
    ● Display the average salary of players for each 'team'.
    ● Plot a bar chart to visualize the average salary of players for each 'team'.
    
#### 10. Extras
    ● Get the index at which the minimum weight value is present.
    ● Sort values based on name in alphabetical order for the rows (the original Dataframe sorting should not change)
    ● Create a series from given dataframe on “name” column and display top and last 10

In [25]:
# Import Libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore") # To ignore unnecessary warnings

#### 1. Load Data:

● Load the nba.csv data into a Pandas DataFrame.

In [28]:
df = pd.read_csv('dataset/nba.csv')
df.head()

Name            Team  Number Position  Age  Weight  \
0  Avery Bradley  Boston Celtics       0       PG   25     180   
1    Jae Crowder  Boston Celtics      99       SF   25     235   
2   John Holland  Boston Celtics      30       SG   27     205   
3    R.J. Hunter  Boston Celtics      28       SG   22     185   
4  Jonas Jerebko  Boston Celtics       8       PF   29     231   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

● Display basic information about the DataFrame.

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 457 entries, 0 to 456
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      457 non-null    object 
 1   Team      457 non-null    object 
 2   Number    457 non-null    int64  
 3   Position  457 non-null    object 
 4   Age       457 non-null    int64  
 5   Weight    457 non-null    int64  
 6   College   373 non-null    object 
 7   Salary    446 non-null    float64
dtypes: float64(1), int64(3), object(4)
memory usage: 28.7+ KB


#### Observations:

1. The dataset contains 457 entries and 8 columns, including 'Name', 'Team', 'Number', 'Position', 'Age', 'Weight', 'College', and 'Salary'.
<br>

2. Some entries have missing values, particularly in the 'College' and 'Salary' columns. There are 373 non-null values in the 'College' column and 446 non-null values in the 'Salary' column.
<br>

3. The data types differ across the columns, with 'Name', 'Team', 'Position', and 'College' being of the object data type, 'Number', 'Age', and 'Weight' being of the int64 data type, and 'Salary' being of the float64 data type.
<br>

4. The dataset presents information related to player attributes such as age, weight, and position, as well as their respective salaries.
<br>

5. The 'Age' and 'Weight' columns are complete, containing 457 non-null entries, indicating that there are no missing values in these columns.
<br>

6. The 'Number' column seems to represent the jersey number of the players in the dataset.
<br>

7. The 'Salary' column has numeric values, indicating the players' salaries, while the 'College' column represents the college where the players attended.

#### 2. Data Cleaning:

● Handle missing values by either removing or imputing them.

In [ ]:
df.

● Remove duplicate rows